src/jit/morph.cpp

```diff
@@ -8038,86 +7804,106 @@ void Compiler::fgMorphTailCallViaHelper(GenTreeCall* call, void* pfnCopyArgs)
         // materialize as embedded stmts in right execution order.
         assert(thisPtr != nullptr);
         call->gtCallArgs = gtNewListNode(thisPtr, call->gtCallArgs);
+        // We changed args so recompute info.
+        call->fgArgInfo = nullptr;
     }
 
-#if defined(_TARGET_AMD64_)
-
-    // Add the extra VSD parameter to arg list in case of VSD calls.
-    // Tail call arg copying thunk will move this extra VSD parameter
-    // to R11 before tail calling VSD stub. See CreateTailCallCopyArgsThunk()
-    // in Stublinkerx86.cpp for more details.
-    if (call->IsVirtualStub())
+    // We may need to pass the target address, for instance for calli.
+    if (help.flags & CORINFO_TAILCALL_STORE_TARGET)
     {
-        GenTree* stubAddrArg = fgGetStubAddrArg(call);
+        GenTree* target;
+        if (call->gtCallType == CT_INDIRECT)
+        {
+            noway_assert(call->gtCallAddr != nullptr);
 
-        // We don't need this arg to be in the normal stub register, so
-        // clear out the register assignment.
-        assert(stubAddrArg->gtRegNum == virtualStubParamInfo->GetReg());
-        stubAddrArg->gtRegNum = REG_NA;
+            // TODO: This should be evaluated last. How?
+            target = call->gtCallAddr;
+            call->gtCallAddr = nullptr;
+        }
+        else
+        {
+            noway_assert(call->gtCallType == CT_USER_FUNC && call->gtCallMethHnd != nullptr);
+            target = fgGetDirectCallTargetAddress(call);
+        }
 
-        // And push the stub address onto the list of arguments
-        call->gtCallArgs = gtNewListNode(stubAddrArg, call->gtCallArgs);
+        call->gtCallArgs = gtNewListNode(target, call->gtCallArgs);
+        call->fgArgInfo = nullptr;
     }
 
-    // Now inject a placeholder for the real call target that Lower phase will generate.
-    GenTree* arg     = gtNewIconNode(0, TYP_I_IMPL);
-    call->gtCallArgs = gtNewListNode(arg, call->gtCallArgs);
+    // This is now a direct call to the store args stub and not a tailcall.
+    call->gtCallType    = CT_USER_FUNC;
+    call->gtCallMethHnd = help.hStoreArgs;
+    call->gtFlags &= ~GTF_CALL_VIRT_KIND_MASK;
+    // TODO: We should be able to clean up this. The tailcall flag is only used
+    // for fast tailcalls now.
+    call->gtCallMoreFlags &=
+        ~(GTF_CALL_M_TAILCALL | GTF_CALL_M_DELEGATE_INV | GTF_CALL_M_SECURE_DELEGATE_INV);
```

Comment:
@dotnet/jit-contrib Would it be better to unconditionally set the flags to 0 (or something else) here? I just ran into a problem with these delegate flags being left set and I am wondering if the flag fields should just be reset entirely.